Combobox: Implement free text search

it-vegard opened the pull request:
### Description

Implement combobox free text search (search inside text, not just from the start)

Fixes #2953
Fixes #2594
Fixes #2223, at least partially. (Not fuzzy-search, but partial search)

### Change summary

- Free text search in Combobox
- Update value on unselecting autocompleted text

it-vegard committed:
Implement fuzzy search in combobox

it-vegard committed:
Merge branch 'main' into feature/combobox-fuzzy-search

KenAJoh commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,16 +141,18 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
           if (activeDecendantId || value) {
             e.preventDefault();
           }
+        } else if (["ArrowLeft", "ArrowUp", "ArrowRight"].includes(e.key)) {
```
Found 2 issues with current implementation:
Combobox now triggers onChange for every ArrowLeft/Right/Up, not only when completing autocomplete.

https://github.com/navikt/aksel/assets/26967723/9e92d6e9-0711-469c-a4dd-e97d03ae02de

Also ArrowUp won't work when using arrowkeys in list, so can only navigate down.
https://github.com/navikt/aksel/assets/26967723/c7c26123-7dc8-4319-b206-d5a5f8ab4903

KenAJoh replied:
Would this be a sufficient guard-clause?

```js
if(value === searchTerm || value === ""){
  // skip onChange 
}
```

KenAJoh replied:
Pushed some changes that hopefully fixes this 🤞

KenAJoh committed:
bug: ArrowUp now works as intended, stopped onChange from running each arrow-press

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filtered-options-util.ts:
```diff
@@ -4,11 +4,16 @@ const normalizeText = (text: string): string =>
+const getValuesStartingWith = (text: string, list: ComboboxOption[]) =>
+  list.find((listItem) =>
+    normalizeText(listItem.label).startsWith(normalizeText(text)),
+  );
```
```suggestion
const getFirstValueStartingWith = (text: string, list: ComboboxOption[]) => {
  const normalizedText = normalizeText(text);
  return list.find((listItem) =>
    normalizeText(listItem.label).startsWith(normalizedText),
  );
};
```

HalvorHaugan commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -106,14 +106,17 @@ const FilteredOptionsProvider = ({
-        `${searchTerm}${filteredOptions[0].label.substring(searchTerm.length)}`,
+        `${searchTerm}${autoCompleteCandidate.substring(searchTerm.length)}`,
       );
       setSearchTerm(searchTerm);
```
This line seems redundant: `setSearchTerm(searchTerm)`

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
Type should be added for param `e` in `handleKeyDown` (line 130).

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,17 +141,24 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
-              toggleIsListOpen(true);
-            }
-            virtualFocus.moveFocusDown();
+          if (e.currentTarget.selectionStart) {
```
This comment seems to be incorrect now

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,17 +141,24 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
           }
+        } else if (["ArrowLeft", "ArrowRight"].includes(e.key)) {
+          if (value !== "" && value !== searchTerm) {
+            onChange(e);
```
Maybe add a comment about why we want to trigger onChange on ArrowLeft, ArrowRight and ArrowUp.

KenAJoh committed:
memo: Added comment

KenAJoh committed:
refactor: Removed extra state-update for searchterm

KenAJoh committed:
Update @navikt/core/react/src/form/combobox/FilteredOptions/filtered-options-util.ts

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
bug: Missed function export after refactor

it-vegard committed:
Add test for filtering in search

it-vegard committed:
Fix comments to better reflect the new code

it-vegard committed:
Change type for onChange to take in only value, not the entire event. This simplifies calling onChange programatically.

it-vegard committed:
Added changesets to reflect the changes in this branch

it-vegard committed:
Merge branch 'main' into feature/combobox-fuzzy-search

it-vegard committed:
KeyboardEvent.currentTarget does not have selectionStart property

it-vegard committed:
Update example

it-vegard committed:
Update storybook examples

it-vegard committed:
Merge branch 'main' into feature/combobox-fuzzy-search

it-vegard committed:
Merge branch 'main' into feature/combobox-fuzzy-search

it-vegard committed:
Add generic to enhance the event type for handleKeyDown

it-vegard committed:
Updated JSDoc

HalvorHaugan commented on aksel.nav.no/website/pages/eksempler/combobox/multi-select-controlled.tsx:
```diff
@@ -38,7 +38,7 @@ const Example = () => {
         filteredOptions={filteredOptions}
         isMultiSelect
-        onChange={(event) => setValue(event?.target.value || "")}
+        onChange={(newValue) => setValue(newValue)}
```
```suggestion
        onChange={setValue}
```

KenAJoh committed:
bug: Fixed undefined value sent to toggleOption callback

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,17 +149,28 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
-            virtualFocus.moveFocusDown();
+          // Reset the value to the search term to cancel autocomplete
+          // if the user moves focus down to the FilteredOptions
+          if (internalRef.current?.selectionStart) {
```
Do we need this check?

it-vegard replied:
I think the purpose was to stop unnecessary calls to setValue, but this does not work, as selectionStart will indicate the position of the cursor if no text is selected. 🤦 

Should we replace it with the following code, or is it harmless to have extra calls to setValue?

```suggestion
          if (value !== "" && value !== searchTerm) {
```

HalvorHaugan replied:
Should be harmless, but might be nice to have the `if` for clearity. Do we need to check that `value` is not empty though?

it-vegard replied:
Good point!
```suggestion
          if (value !== searchTerm) {
```

KenAJoh committed:
revert: Removed test for what was corret interaction

it-vegard committed:
Merge branch 'main' into feature/combobox-fuzzy-search

# Conflicts:
#	@navikt/core/react/src/form/combobox/combobox.stories.tsx

it-vegard committed:
Changed method signatures for onChange earlier, but forgot these two tests

it-vegard committed:
Add test for triggering onChange on every typed character in Combobox with autocomplete.

Currently fails on the last character, as it does not actually change the word in the input field.

it-vegard committed:
Move onChangeHandler from triggering onChange to onInput, as the former does not trigger when typing the last character of an autocompleted word.
onInput seems to work more consistently, but will test multiple use cases.

it-vegard committed:
Update aksel.nav.no/website/pages/eksempler/combobox/multi-select-controlled.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/react/src/form/combobox/Input/Input.tsx

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -184,10 +215,11 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
         value={value}
         onBlur={() => virtualFocus.moveFocusToTop()}
-        onChange={onChangeHandler}
+        onClick={() => value && updateInputValue(value)}
```
`externalOnChange()` is not being called if you "complete" the selection by clicking. Is that intended?
(I.e. should we call `onChange` instead of `updateInputValue` here?)

it-vegard replied:
Enig. Går for følgende:
```suggestion
        onClick={() => value && onChange(value)}
```

HalvorHaugan commented on .changeset/wild-owls-tan.md:
```diff
@@ -0,0 +1,6 @@
+"@navikt/ds-css": minor
+---
+
+Improved search in combobox - find hits anywhere in the label
```
```suggestion
Improved search in Combobox - find hits anywhere in the label
```

it-vegard committed:
Update .changeset/wild-owls-tan.md

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Update @navikt/core/react/src/form/combobox/Input/Input.tsx

it-vegard committed:
updateInputValue is no longer neccessary

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -184,10 +214,11 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
         value={value}
         onBlur={() => virtualFocus.moveFocusToTop()}
-        onChange={onChangeHandler}
+        onClick={() => value && onChange(value)}
```
```suggestion
        onClick={() => value !== searchTerm && onChange(value)}
```

HalvorHaugan committed:
Update @navikt/core/react/src/form/combobox/Input/Input.tsx
